llvm/include/llvm/DebugInfo/Symbolize/SymbolizableObjectFile.h

```diff
@@ -82,12 +82,8 @@ class SymbolizableObjectFile : public SymbolizableModule {
     // Non-zero if this is an ELF local symbol. See the comment in
     // getNameFromSymbolTable.
     uint32_t ELFLocalSymIdx;
-
-    bool operator<(const SymbolDesc &RHS) const {
-      return Addr != RHS.Addr ? Addr < RHS.Addr : Size < RHS.Size;
-    }
   };
-  std::vector<SymbolDesc> Symbols;
+  std::map<uint64_t, SymbolDesc> Symbols;
```

Comment:
Are you sure you want a `std::map` and not a sorted vector. https://llvm.org/docs/ProgrammersManual.html#map-like-containers-std-map-densemap-etc gives a lot of thoughts on this and provides various other alternatives too that might be more efficient.

Reply:
I think `map` is better than [sorted vector](https://llvm.org/docs/ProgrammersManual.html#a-sorted-vector) for my case. In https://github.com/llvm/llvm-project/pull/69553, when add a new symbol, I need to know if there is any existing symbol that has same address. If I am understanding sorted vector right, each time when a new symbol is inserted to a sorted vector, it has to be sorted manually and then for a new insertion, use a log(n) query(std::lower_bound) to check if the address already exists. I don't see the benefit compared with `map`.

Reply:
The main benefit is the performance. As explained in the programming guide, `std::map` is not a very efficient data structure. Please do some runtime comparisons of the tool with and without this change, where you have for both a large number of elements and a small number. In particular, how does the runtime and memory usage compare before and after the change?

>  If I am understanding sorted vector right, each time when a new symbol is inserted to a sorted vector, it has to be sorted manually

The programming guide seems pretty clear in this regards: a sorted vector is the right approach if you want to insert lots of elements but don't care about the order during insertion. In other words, you haven't understood "sorted vector" correctly. The existing usage looks precisely like is described in the programmer's manual: lots of insertions, followed by a sort, and only then is the vector looked in:

> This approach works really well if your usage pattern has these two distinct phases (insert then query), and can be coupled with a good choice of [sequential container](https://llvm.org/docs/ProgrammersManual.html#ds-sequential).

> when add a new symbol, I need to know if there is any existing symbol that has same address.

Why do you need to know this _at insertion time_ rather than just sorting things differently?

Reply:
> Why do you need to know this at insertion time rather than just sorting things differently?

Because of the data we stored to `Symbols`. For now, The symbol is stored as:
```
  struct SymbolDesc {
    uint64_t Addr;
    // If size is 0, assume that symbol occupies the whole memory range up to
    // the following symbol.
    uint64_t Size;

    StringRef Name;
    // Non-zero if this is an ELF local symbol. See the comment in
    // getNameFromSymbolTable.
    uint32_t ELFLocalSymIdx;

    bool operator<(const SymbolDesc &RHS) const {
      return Addr != RHS.Addr ? Addr < RHS.Addr : Size < RHS.Size;
    }
  };
```

When we do the sorting, we need to know some symbol flags besides address and size for XCOFF. See the comments here https://github.com/llvm/llvm-project/pull/69553/files#diff-795c0eb20ded2ff47056aa629fc0e183c20225b8fd40b607627ab3106995779dR222-R225

Is it acceptable to you that we add a new element (maybe a `SymbolRef` type and can eliminate current field `ELFLocalSymIdx`) in `struct SymbolDesc`? If using this way, we don't need to change the container, can still use vector and do the sort once.